adapter/service/scenectrl/service.go

```diff
@@ -0,0 +1,124 @@
+package scenectrl
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/futurehomeno/fimpgo"
+	"github.com/futurehomeno/fimpgo/fimptype"
+
+	"github.com/futurehomeno/cliffhanger/adapter"
+	"github.com/futurehomeno/cliffhanger/adapter/cache"
+)
+
+const (
+	PropertySupportedScenes = "sup_scenes"
+)
+
+// DefaultReportingStrategy is the default strategy used by the service for periodic reports.
+var DefaultReportingStrategy = cache.ReportOnChangeOnly()
+
+// Controller is an interface representing an actual device.
+// In a polling scenario implementation might require some safeguards against excessive polling.
+type Controller interface {
+	// SetSceneCtrlScene sets the scene of the device.
+	SetSceneCtrlScene(scene string) error
+	// SceneCtrlReport returns a current scene value.
+	SceneCtrlSceneReport() (string, error)
+}
+
+// Service is an interface representing a presence FIMP service.
+type Service interface {
+	adapter.Service
+
+	// SetScene sets the scene of the device.
+	SetScene(scene string) error
+	// SendSceneReport sends a scene report. Returns true if a report was sent.
+	// Depending on a caching and reporting configuration the service might decide to skip a report.
+	// To make sure report is being sent regardless of circumstances set the force argument to true.
+	SendSceneReport(force bool) (bool, error)
+}
+
+// Config represents a service configuration.
+type Config struct {
+	Specification     *fimptype.Service
+	Controller        Controller
+	ReportingStrategy cache.ReportingStrategy
+}
+
+// NewService creates a new instance of a presence FIMP service.
+func NewService(
+	mqtt *fimpgo.MqttTransport,
+	cfg *Config,
+) Service {
+	cfg.Specification.EnsureInterfaces(requiredInterfaces()...)
+
+	if cfg.ReportingStrategy == nil {
+		cfg.ReportingStrategy = DefaultReportingStrategy
+	}
+
+	return &service{
+		Service:           adapter.NewService(mqtt, cfg.Specification),
+		controller:        cfg.Controller,
+		lock:              &sync.Mutex{},
+		reportingStrategy: cfg.ReportingStrategy,
+		reportingCache:    cache.NewReportingCache(),
+	}
+}
+
+// service is a private implementation of a presence FIMP service.
+type service struct {
+	adapter.Service
+
+	controller        Controller
+	lock              *sync.Mutex
+	reportingCache    cache.ReportingCache
+	reportingStrategy cache.ReportingStrategy
+}
+
+// SetScene sets the scene of the device.
+func (s *service) SetScene(scene string) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if err := s.controller.SetSceneCtrlScene(scene); err != nil {
```

Comment:
Before passing a `scene` to the controller, I'd check if provided scene is in `sup_scenes` and throw an error if not. Please remember about the tests for that case.

Reply:
Fixed